### ISSUE TITLE
Refactor members table: group entries by MemberCollectiveId

### ISSUE THREAD

**Issue by Betree** [opencollective/opencollective#3941]
# Problem

In the current model, we add one `Member` entry per MemberCollectiveId/Role/TierId. So if you contribute to 2 tiers and you're an admin of a collective, it will be represented in the DB with 3 `Members` entry like:
| id    | MemberCollectiveId | CollectiveId | TierId | role      | 
|-------|----------------------|----------------|----------|-------------| 
| 1 | 42                | 10          | 50         | BACKER | 
| 2 | 42                | 10          | 51     | BACKER      | 
| 3 | 42                | 10            |          | ADMIN | 

There are a few problems with this approach:
- In many [places](https://github.com/opencollective/opencollective-frontend/blob/7722008a3c9422f23d160fc529339543745802a6/components/collective-page/sections/Contributions.js#L165) in the frontend we have to manually group the members to make sure we don't display duplicates. This doesn't play well with pagination, and impacts performance.
- The current workaround in the API is to use the `contributors` endpoint which basically does what's proposed in this issue through an SQL request with a `GROUP BY` on the `MemberCollectiveId`. It works, but performances are not great and we have to use intensive caching to make that work.
- We had the case of people getting confused by our API because they expected a contributor to belong to a different tier/role (turns out they were just looking at the wrong member entry).

**To summarize**, the goal of this issue is to improve performance and simplify our API.

# Proposal

I propose that we merge all the lines in `Members` by `MemberCollectiveId`<>`CollectiveId` pair (we can eventually add a unique index on it to enforce that). Two columns are going to be in our way to achieve that:
- `role`: should be replaced with `roles`, an array of enum (`ADMIN`, `BACKER`, etc.)
- `TierId`: should be replaced with `TiersIds`, an array of linked tiers

| id    | MemberCollectiveId | CollectiveId | TiersIds | roles      | 
|-------|----------------------|----------------|----------|-------------| 
| 1 | 42                | 10          | [50, 51]         | [BACKER, ADMIN] | 

An alternative is to get rid of the `roles` column entirely, and rather have specific columns for roles:
| id    | MemberCollectiveId | CollectiveId | TiersIds | isAdmin      | isFinancialContributor
|-------|----------------------|----------------|----------|-------------|-------------| 
| 1 | 42                | 10          | [50, 51]         | true | true

We could also store the tier memberships in an entirely new table, like `MemberTiers`:

There may be a lot of code to update to support that change properly, so complexity should be better estimated before starting anything.

**Comment by sbinlondon** [comment]
Thanks for explaining the problem with the current approach. I agree that this could help performance and also contributors to our codebase/users of the API understand stuff better. 👍

**Comment by Betree** [comment]
Changing the design of the `Members` as proposed in this issue is a gigantic task that's unlikely to happen anytime soon.

I now believe materialized views can provide a better alternative: by creating a `CollectiveContributors` view, we'd be able to implement the columns described in this issue and get the same performance benefits without having to worry about breaking existing code. Going further, using an [eager materialized view](https://hashrocket.com/blog/posts/materialized-view-strategies-using-postgresql#eager-materialized-view) would guarantee fresh data and quick updates.

Basically:
- Create a materialized view by adapting [this query](https://github.com/opencollective/opencollective-api/blob/200ecd65dbe40e610503c4ffa7a5459f43c12a45/server/lib/contributors.ts#L90)
- Add Postgres trigger to update the view for an account whenever members are added/removed/edited
- Update the `contributors` field to use this view
